### PR TITLE
Add set_name and clear_attributes functions to BytesStart

### LINKED ISSUE #148: Re-using BytesStart byte vector
Currently, `BytesStart` will [allocate a `Vec<u8>`](https://docs.rs/quick-xml/0.13.2/src/quick_xml/events/mod.rs.html#164-172) the first time you push additional attributes for the element. That's unfortunate if you're emitting lots of `BytesStart` in a tight loop. It would be better if it was possible to re-use a `BytesStart` (specifically its internal `Cow::Owned`) across iterations. I _think_ all that'd be needed would be two methods: `set_name` and `clear_attributes`:
```rust
impl BytesStart {
    fn set_name(&mut self, name: &[u8]) {
        let bytes = self.buf.to_mut();
        bytes.clear();
        bytes.extend(name);
        self.name_len = name.len();
    }

    fn clear_attributes(&mut self) {
        self.buf.to_mut().truncate(self.name_len);
    }
}
```

### PR DESCRIPTION
**Pull request by meven**
Fixes #148

### CODE REVIEW

**Comment by jonhoo** [comment]
Those `fn`s should be `pub fn`, that's why you're getting the warnings :)

**Comment by meven** [comment]
> Those `fn`s should be `pub fn`, that's why you're getting the warnings :)

Thanks ;) that's been taking care of.

**Comment by jonhoo** [comment]
@tafia looks like the Travis failure is just `rustfmt-preview` being unavailable for nightly (the component is now just called `rustfmt`).

File: src/events/mod.rs [hunk 1]
```diff
@@ -170,6 +170,23 @@ impl<'a> BytesStart<'a> {
         bytes.extend_from_slice(&*a.value);
         bytes.push(b'"');
     }
+
+    /// Edit the name of the BytesStart in-place
```
Comment (tafia) [inline src/events/mod.rs]: Can you add a warning stating that the content of the `BytesStart` will be cleared?
Someone may not expect attributes to disappear.

**Comment by tafia** [comment]
Thanks for the PR and sorry for the late response.
Can you please just modify the comment on `set_name`?
I'll push changes to travis in the meantime.

File: src/events/mod.rs (hunk 1 above)
Comment (meven) [reply src/events/mod.rs]: Good catch, I have added a warning.

**Comment by meven** [comment]
Last change made me think that perhaps `set_names` should not edit the attributes and keep them.
It would end up as :
```rust
    /// Edit the name of the BytesStart in-place
    pub fn set_name(&mut self, name: &[u8]) -> &mut BytesStart<'a> {
        {
            let bytes = self.buf.to_mut();
            bytes.splice(..self.name_len, name.to_vec());
        }
        self.name_len = name.len();
        self
    }
```

**Comment by tafia** [comment]
Sounds better indeed!

**Comment by meven** [comment]
> Sounds better indeed!

Done, I feel like this can be merged

**Comment by tafia** [comment]
Thanks!

File: src/events/mod.rs [hunk 2]
```diff
@@ -170,6 +170,22 @@ impl<'a> BytesStart<'a> {
...
+    pub fn set_name(&mut self, name: &[u8]) -> &mut BytesStart<'a> {
+        {
+            let bytes = self.buf.to_mut();
+            bytes.splice(..self.name_len, name.to_vec());
```
Comment (jonhoo) [inline src/events/mod.rs]: @meven  I don't think the `to_vec` call here is needed. It'd be good to avoid the extra allocation!

Comment (meven) [reply src/events/mod.rs]: I did not find a better way I would expect the same but I ran into 
```console
error[E0271]: type mismatch resolving `<&[u8] as std::iter::IntoIterator>::Item == u8`
   --> src/events/mod.rs:178:19
    |
178 |             bytes.splice(..self.name_len, name);
    |                   ^^^^^^ expected reference, found u8
    |
    = note: expected type `&u8`
               found type `u8`
```
It seems I still need some conversion, name `&[u8]` implements `IntoIterator<u8>` but I need an `IntoIterator<&u8>`. Maybe There is some simple trick I am missing.

Comment (jonhoo) [reply src/events/mod.rs]: Ah, I've run into this in the past. It's a little silly. Basically what you want is `name.iter().cloned()` or `name.iter().map(|&b| b)` to tell the compiler that you want to utilize the fact that `u8` is `Copy`.

Comment (meven) [reply src/events/mod.rs]: Thanks, will make another PR

**Pull request merged**